### PR TITLE
fix dtype/device property not getting updated in submodules

### LINKED ISSUE #2442: validation_epoch_end needs to return CUDA tensors
## 🐛 Bug

I'm not sure if this is expected behaviour or not, but upgrading to the latest version (from 0.8.1) caused my `validation_epoch_end` to break. It appears that a CUDA tensor is expected for the metric where before the tensor was device agnostic.

This was using sklearn's `roc_auc_score`. I haven't yet got around to testing PL's new metrics.

Feel free to close if this is expected behaviour.

### To Reproduce

This is my `validation_epoch_end`. Uncommenting `.to(avg_loss.device)` allows this to run with the dev version of PL.

This was run with `ddp`, `precision=16` using Apex AMP.

```
def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
        y_pred = torch.cat([x["y_pred"].squeeze() for x in outputs]).cpu().numpy()
        y_true = torch.cat([x["y_true"].squeeze() for x in outputs]).cpu().numpy()

        metric = torch.tensor(roc_auc_score(y_true, y_pred))  # .to(avg_loss.device)

        tensorboard_logs = {
            "loss/validation": avg_loss,
            "auc": metric,
        }
        return {"val_loss": avg_loss, "log": tensorboard_logs, "auc": metric}
```

The error message can be seen here: https://github.com/PyTorchLightning/pytorch-lightning/issues/2411#issuecomment-652171874 

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version: 3.8.2
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: Dual GPU (ddp)
 - Any other relevant information: Apex AMP

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #2442 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

The new tests fail on master and demonstrate the bug.

### CODE REVIEW

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 1]
```diff
@@ -153,5 +145,17 @@ def half(self) -> Module:
...
+
+        def apply_fn(module):
+            if not isinstance(module, DeviceDtypeModuleMixin):
+                return
+            if device is not None:
+                module._device = device
+            if dtype is not None:
+                module._dtype = dtype
+
+        self.apply(apply_fn)
```
Comment (Borda) [inline pytorch_lightning/utilities/device_dtype_mixin.py:161]: what is the reason of using `apply`?

**Comment by williamFalcon** [comment]
@awaelchli can you also test this with dp and ddp_spawn backends?
make sure to use two different functions and not a parametrized test or it won’t work

**Comment by awaelchli** [comment]
@justusschock Yes, I understand. It took me a while to figure it out. The thing is, the super calls the _apply, and if you look inside you will see that will only dispatch the .cuda, .to, etc. to buffers and parameters! But we need our special properties to be set on the module, so we need to call apply (not _apply).
